### PR TITLE
Bundle assemblies at 4K for linux arm64

### PR DESCRIPTION
**Pull request by sbomer**
On ARM64 Linux, single-file apps were segfaulting because of an incorrect address computed in R2R code. See https://github.com/dotnet/runtime/issues/41832 for details about the issue.

This is a temporary fix that uses 4K offsets within the bundle. This will preserve the page alignment when we mmap assemblies, working around the lack of fixups for `adrp`; `add` instruction sequences.

I've confirmed that the fix, together with these SDK changes, produces a functional single-file app on Linux ARM64: https://github.com/sbomer/sdk/commit/19998e630c7aeb46e19898d4c2fb9b8df57a6ee7. I will make the SDK changes in the rc2 branch after this fix is ported and flows to the SDK.

### CODE REVIEW

**Review by swaroop-sridhar — APPROVED** [review]

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs
Comment (swaroop-sridhar) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs:55]: Please add the information to the design doc too, in [this section](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#peimage-loader). Add a Linux-ARM64 limitation, in addition to Windows limitations.

Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs]: Thanks! I filed https://github.com/dotnet/runtime/issues/41832, updated the comment, and am updating the docs at https://github.com/dotnet/designs/pull/156.

Comment (swaroop-sridhar) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs]: Thank you very much, looks good.

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs
Comment (agocke) [inline src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:32]: Don't we want to break the SDK build? That is, don't we want the insertion that brings this into the SDK to be accompanied with our change on the SDK side?

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs
Comment (agocke) [inline src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs:28]: Do we need to store the AssemblyAlignment in a field? It looks like it could be a calculated property based on the Arch. Moreover, do we need to store it on `TargetInfo` at all? It's not really a property of the target info. It was previously stored in Bundler.cs, which seems like the only place it's needed.

Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs:28]: I made it a field for consistency with `DefaultOptions` and `BundleVersion`, but don't have a strong preference. To me it seems like it belongs with `TargetInfo` because it depends on the alignment requirements of the target platform - for example if the bundler accepted an option to change the alignment, we would still only allow multiples of this value (similar to `DefaultOptions`).

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs
Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:32]: It would block dependency flow from runtime -> SDK until that change went in, which might be nice to avoid in case there are other fixes going through. If temporarily blocking it is ok, I'll remove this.

Comment (agocke) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:32]: It will only block ingestion of 6.0 bits right? That seems fine to me.

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs
Comment (agocke) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs:28]: > belongs with TargetInfo because it depends on the alignment requirements of the target platform

I would disagree -- `TargetInfo` is a collection of data about the target, while alignment is a piece of data about the bundler, which is dependent on `TargetInfo`. My general strategy would be for TargetInfo to carry only the essentials of the target, because that makes it more reusable and lighter weight for dependent components.

However I think this is all pretty academic because all of this code is pretty small, so it doesn't particularly matter to me either way 😄

**Review by agocke — APPROVED** [review]
Aside from overload LGTM

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs
Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: Removed this as discussed. The plan is to make the SDK change quickly so that the dependency flow is not blocked for long.

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs
Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/TargetInfo.cs:28]: Point taken - I'll leave it this way for consistency, but we could consider removing bundler-related decisions from it.

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs
Comment (swaroop-sridhar) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: We also need the change migrated to 5.0 branches in both repos, isn't it? Can we briefly take breaking changes there too? 
In SDK, it is typical to only checkin to the 5.0 branch release, and let the change flow into master branch.

Comment (agocke) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: Not sure I understand what's breaking about this change, aside from the fact that the SDK needs a parallel fix -- but it's not like the SDK can not take that fix, fixing the bug means changing both repos.

Comment (swaroop-sridhar) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: The SDK's invocation of the bundler needs to pass in the architecture argument, but it can only do so when this change flows into the SDK through HostModel version update. Given the position of `Architecture` argument, this looks like it'll break the SDK integration job; Isn't it?. So, we have the following options: 
* Make the update as part of SDK integration commit
* Make the update in two steps -- once SDK gets a chance to update.

In this case, the change is small, and its OK to follow the former approach. But, we've followed the latter approach so far for HostModel/SDK interdependencies (originally per Nick's recommendation).

Comment (agocke) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: I'm not sure why he suggested (2), but I rather prefer (1), as it removes the possibility of accidentally dropping the corresponding change somewhere.

Comment (swaroop-sridhar) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: Nick suggested (2) because the SDK parts of the change are typically non-trivial, and they could be independently reviewed in a normal PR, rather than in the automated integration job, and because they can be independently reverted if there's a problem without affecting the other parts of integration.

But in this case, the change on the SDK side is very small, so I'm fine with (1).

File: src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs
Comment (vitek-karas) [inline src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs:10]: Nit: sort

File: src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs
Comment (agocke) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: That makes sense. For larger changes (2) could definitely be appropriate, especially if we're concerned that CI wouldn't fully validate the change.

Comment (sbomer) [reply src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs]: Just to be clear - I do plan to get the change reviewed in a normal PR, not in the automated job. I'll include a manual dependency update in my PR, but the automated flow will be blocked until my PR is merged.

File: src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs
Comment (vitek-karas) [inline src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs:114]: This is basically wrong - for example `win81-arm64-aot` is a valid RID (see https://github.com/dotnet/runtime/blob/master/src/libraries/pkg/Microsoft.NETCore.Platforms/runtime.json and scroll to the end of the file).:
* Since this is test-only code - we should take the second part after splitting on `-`

**Review by vitek-karas — APPROVED** [review]

File: src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs
Comment (sbomer) [reply src/installer/tests/Microsoft.NET.HostModel.Tests/Helpers/BundleHelper.cs]: There's also `linux-musl-arm64` - I'll fix it to do what the SDK does.

**Review by vitek-karas — APPROVED** [review]

**Pull request merged**
